File: android-contract/src/main/kotlin/com/audiochoice/contracts/NarrationContracts.kt
package com.audiochoice.contracts

import kotlinx.serialization.Serializable

/**
 * Asks the server to find filterable content in a book's text, for a book with no audiobook.
 *
 * Reuses [BookFingerprint] rather than declaring a narration-specific identity. A book is the
 * same book whether it arrived as audio or as an EPUB, and sharing the fingerprint is what
 * lets one listener's scan serve the next listener who imports the same file.
 *
 * [bookText] is the only field carrying the book itself, and the server holds it for the
 * length of one request. Nothing sends it twice: a scan is looked up by fingerprint first.
 */
@Serializable
data class NarrationTextScanRequest(
    val fingerprint: BookFingerprint,
    val bookText: String,
    val language: String? = null,
) {
    /**
     * Prints the text's length rather than the text.
     *
     * A data class prints every property, so one log line holding the request would write a
     * whole novel into logcat. That is the same disclosure the feature is careful to avoid,
     * arriving by accident, so the safe rendering is the default rather than something each
     * call site has to remember.
     */
    override fun toString(): String =
        "NarrationTextScanRequest(sha256=${fingerprint.sha256}, " +
            "bookTextCharacters=${bookText.length}, language=$language)"
}

/**
 * Filter events for a book's text.
 *
 * The offsets in each [ScanEvent] are **character offsets into the book's text**, carried in
 * the same `startTime`/`endTime` fields an audio scan uses for seconds. That reuse is
 * deliberate: it lets the entire existing filter stack -- the category switches, the
 * per-word profanity grouping, the aggregate descriptions -- work on a narrated book without
 * a parallel implementation. It is also the single most dangerous thing about this contract,
 * because a value from here reaching a seek call would move the listener by tens of hours.
 * Anything reading these events must keep them apart from playback positions.
 *
 * [bookTextCharacters] is what makes a stale scan detectable: if the book's text is no longer
 * this length, the offsets no longer index it, and the events must be discarded rather than
 * applied to the wrong passages.
 */
@Serializable
data class NarrationTextScanResponse(
    val events: List<ScanEvent> = emptyList(),
    val scanDate: String? = null,
    val scannerVersion: String = "",
    val taxonomyVersion: String = "",
    val bookTextCharacters: Int = 0,
)

File: backend/AudioChoice.Api/Contracts/NarrationContracts.cs
namespace AudioChoice.Api.Contracts;

/// <summary>
/// Values shared by the audio and text scanning paths, so neither can drift from the other
/// by way of a copied literal.
/// </summary>
public static class ScanContracts
{
    /// <summary>
    /// The content taxonomy the events were drawn from.
    /// </summary>
    /// <remarks>
    /// Must match <see cref="CloudScanResponse"/>'s default. A narrated book's filter
    /// switches are the same switches an audiobook's are, and they are matched by category
    /// and group identifier, so a taxonomy disagreement would present a listener with
    /// toggles that control nothing. A contract test asserts the two agree.
    /// </remarks>
    public const string TaxonomyVersion = "2.0";
}

/// <summary>
/// A request to find filterable content in a book's text, for a book with no audiobook.
/// </summary>
public sealed record NarrationTextScanRequest(
    BookFingerprint Fingerprint,
    string BookText,
    string? Language = null)
{
    /// <summary>
    /// Prints the text's length instead of the text.
    /// </summary>
    /// <remarks>
    /// The generated <c>ToString</c> for a record prints every member, so one logger call
    /// that happens to include the request -- a log scope, an unhandled-exception dump, a
    /// developer adding <c>{Request}</c> to a message -- would write an entire novel into a
    /// log sink that is retained, shipped and searchable. That is the same disclosure the
    /// endpoint exists to avoid, arriving through the back door, so the safe rendering is
    /// the default one rather than something each call site must remember.
    /// </remarks>
    public override string ToString() =>
        $"NarrationTextScanRequest {{ Sha256 = {Fingerprint?.Sha256}, " +
        $"BookTextCharacters = {BookText?.Length ?? 0}, Language = {Language} }}";
}

/// <summary>
/// Filter events for a book's text, in character offsets into that text.
/// </summary>
/// <remarks>
/// Carries no part of the text back. The offsets say where a passage is, and the client
/// already holds the book, so returning the words would add nothing but exposure.
/// </remarks>
public sealed record NarrationTextScanResponse(
    IReadOnlyList<ScanEvent> Events,
    DateTimeOffset ScanDate,
    string ScannerVersion,
    string TaxonomyVersion,
    int BookTextCharacters);

/// <summary>
/// A completed text scan, as the pipeline produces it and the store records it.
/// </summary>
/// <remarks>
/// <see cref="BookTextCharacters"/> is kept because it is what makes an out-of-range event
/// offset detectable later without keeping the text it indexes.
/// </remarks>
public sealed record NarrationTextScan(
    IReadOnlyList<ScanEvent> Events,
    DateTimeOffset ScanDate,
    string ScannerVersion,
    string TaxonomyVersion,
    int BookTextCharacters,
    string? Language)
{
    public NarrationTextScanResponse ToResponse() =>
        new(Events, ScanDate, ScannerVersion, TaxonomyVersion, BookTextCharacters);
}

/// <summary>
/// The listener's agreement to premium synthesis, which sends a chapter's text off the device.
/// </summary>
public sealed record NarrationAgreement(
    string Version,
    string Text);

/// <summary>Voices a listener may choose, with the agreement that premium requires.</summary>
public sealed record NarrationVoicesResponse(
    IReadOnlyList<NarrationVoiceDescriptor> Voices,
    string AgreementVersion,
    string AgreementText);

/// <summary>One voice, with a fixed pre-rendered sample rather than one made on demand.</summary>
public sealed record NarrationVoiceDescriptor(
    string VoiceID,
    string DisplayName,
    string Language,
    string Provider,
    string SampleUrl);

/// <summary>Records that a listener accepted the premium synthesis agreement.</summary>
public sealed record NarrationAcknowledgementRequest(
    string AgreementVersion,
    string AgreementText);

public sealed record NarrationAcknowledgementResponse(
    string AgreementVersion,
    DateTimeOffset AcceptedAt);

/// <summary>
/// Asks for one chapter to be spoken.
/// </summary>
/// <remarks>
/// Units arrive with filtered characters already removed, so nothing the listener asked to have
/// filtered is ever sent. The <c>ToString</c> override is what keeps a log scope from printing
/// the chapter.
/// </remarks>
public sealed record NarrationChapterRequest(
    BookFingerprint Fingerprint,
    int ChapterIndex,
    string VoiceID,
    string? Language,
    IReadOnlyList<NarrationUnitRequest> Units)
{
    public int CharacterCount => Units.Sum(unit => unit.Text.Length);

    public override string ToString() =>
        $"NarrationChapterRequest {{ Sha256 = {Fingerprint?.Sha256}, " +
        $"ChapterIndex = {ChapterIndex}, VoiceID = {VoiceID}, " +
        $"Units = {Units.Count}, Characters = {CharacterCount} }}";
}

public sealed record NarrationUnitRequest(
    int StartCharacter,
    int EndCharacter,
    string Text);

public sealed record NarrationChapterAccepted(Guid JobID, string Status);

/// <summary>
/// A chapter job's state, and its audio once there is any.
/// </summary>
/// <remarks>
/// <see cref="AudioBase64"/> carries the finished chapter. Returned in the response rather than
/// through a signed URL to a storage container: chapter audio is derived closely enough from a
/// listener's book that keeping it in cloud storage would be the same disclosure the text
/// handling is careful to avoid, and it would sit there under a retention policy rather than
/// under the absence of anywhere to put it.
/// </remarks>
public sealed record NarrationChapterStatus(
    Guid JobID,
    int ChapterIndex,
    string Status,
    string? Provider = null,
    string? ModelVersion = null,
    string? VoiceID = null,
    double DurationSeconds = 0,
    IReadOnlyList<NarrationUnitTiming>? Timings = null,
    string? AudioBase64 = null,
    string? Error = null)
{
    public override string ToString() =>
        $"NarrationChapterStatus {{ JobID = {JobID}, ChapterIndex = {ChapterIndex}, " +
        $"Status = {Status}, Provider = {Provider}, DurationSeconds = {DurationSeconds:F1}, " +
        $"Timings = {Timings?.Count ?? 0}, AudioBase64Length = {AudioBase64?.Length ?? 0} }}";
}

/// <summary>Where one unit's audio sits, measured from the chapter's own start.</summary>
public sealed record NarrationUnitTiming(
    int StartCharacter,
    int EndCharacter,
    double StartSeconds,
    double EndSeconds);
